### PR TITLE
Fix F821 and flake8

### LINKED ISSUE #15055: pre-commit check cannot see missing import
pre-commit on the PR was green, and my editor hooked up to flake8 did not catch it anymore (it used to). What is going on here? This sounds like a bug in the pre-commit/ruff checks. I feel like we went overboard with such checks, making them so complicated that it is starting to fail because half the devs don't know how to read all the settings.

`E           NameError: name 'nullcontext' is not defined`

(The problem above is I forgot to add `from contextlib import nullcontext` but it was not caught until CI ran.)

@nstarman , @eerovaher , or @WilliamJamieson , do you know how to fix this in the settings?

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

It was noticed that we were not catching undefined names with our ruff linting configuration. This PR enables this check and fixes the issues uncovered by this.

Additionally, it was noticed that our flake8 configuration does not catch this error and is finding additional errors due to the various portions of our code excluded from black. This PR also resolves this issue.

Fixes #15055

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -164,12 +164,33 @@ doctest_subpackage_requires =
...
+    astropy/units/photometric.py:F821
+    astropy/units/quantity.py:E266
+    astropy/units/si.py:F821
+    astropy/units/tests/test_quantity_decorator.py:F821
+    astropy/units/tests/test_quantity_non_ufuncs.py:E201,E241,F821
+    astropy/units/tests/test_quantity_typing.py:F821
+    astropy/wcs/wcs.py:F821
+    astropy/wcs/wcsapi/fitswcs.py:F821
+    astropy/wcs/wcsapi/utils.py:E127,E128
+    docs/conf.py:F401
```
Comment (pllim) [inline setup.cfg:193]: I don't grok this, why all these extra lines with code besides F821? I will just have to trust you on this one.

**Review by pllim — APPROVED** [review]
Thanks!

File: setup.cfg (hunk 1 above)
Comment (WilliamJamieson) [reply setup.cfg:193]: I fixed all the flake8 errors, there were more than just the F821.

File: astropy/visualization/wcsaxes/ticklabels.py [hunk 2]
```diff
@@ -8,6 +8,8 @@
 
 from .frame import RectangularFrame
 
+from astropy.utils.exceptions import AstropyDeprecationWarning
```
[the remark this answers is not in the document]
Comment (WilliamJamieson) [reply astropy/visualization/wcsaxes/ticklabels.py:11]: The warning is not tested at all.

**Comment by pllim** [comment]
The CI failure is some URL timeout error from IERS, so we can ignore. Thanks!

**Comment by pllim** [comment]
@WilliamJamieson , are you interested to manually backport this? I think at least the missing import should be backported because it is a bug?

**Comment by WilliamJamieson** [comment]
Sure, I can do this.

**Pull request merged**